core/utils.ts

```diff
@@ -116,20 +116,28 @@ export function validate(compiledGraph: dataform.ICompiledGraph): dataform.IGrap
   allActions.forEach(action => {
     if (allActions.filter(subAction => subAction.name == action.name).length > 1) {
       const actionName = action.name;
-      const message = `Duplicate action name detected, names must be unique across tables, assertions, and operations: "${action.name}"`;
+      const message = `Duplicate action name detected. Names within a schema must be unique across tables, assertions, and operations: "${
+        action.name
+      }"`;
       validationErrors.push(dataform.ValidationError.create({ message, actionName }));
     }
   });
 
   const actionsByName: { [name: string]: dataform.IExecutionAction } = {};
   allActions.forEach(action => (actionsByName[action.name] = action));
 
-  // Check all dependencies actually exist.
+  // Check all dependencies actually exist and are not ambiguous.
   allActions.forEach(action => {
     const actionName = action.name;
     (action.dependencies || []).forEach(dependency => {
-      if (allActionNames.indexOf(dependency) < 0) {
-        const message = `Missing dependency detected: Node "${action.name}" depends on "${dependency}" which does not exist.`;
+      const [matchedDep, err] = matchFQName(dependency, allActionNames);
```

Comment:
if we move dependency resolution into `compile()` functions as discussed above, we shouldn't need this code (i believe!).

Reply:
That is right but then we need to change the validation errors into compilation errors and change the test cases accordingly. And it is not only this one but also circular dependencies, etc. Since this change is already quite large, I didn't want to mess too many things up. Let's do that on Issue https://github.com/dataform-co/dataform/issues/247 once and for all.